### DISCUSSION TITLE
Impossible to type hint `numpy.arange` in strict mode?

### DISCUSSION THREAD

**Discussion by wolph** [microsoft/pyright#5204] [Q&A] [answered]
Note that this might be a bug in `numpy`, but it seems like a bug in `pyright` as well since even casting makes no difference.

I've tried every type of hint and cast I can think of (`np.ndarray` with a `np.dtype`, `npt.NDArray` with `npt.Dtype`, etc...) both with and without generic arguments and as generic arguments I've tried regular ints, floats, numpy ints and floats of varying sizes, etc....

The simplest example, but you can also try and assign it to a variable, it makes no difference:
```
import typing
import numpy

typing.reveal_type(numpy.arange(5))
```

When running pyright in strict mode this gives me:
```
error: Type of "arange" is partially unknown
    Type of "arange" is "Overload[(stop: bool | bool_ | int | integer[Unknown], /, *, dtype: None = ..., like: _SupportsArrayFunc | None = ...) -> ndarray[Any, dtype[signedinteger[Any]]], (start: bool | bool_ | int | integer[Unknown], stop: bool | bool_ | int | integer[Unknown], step: bool | bool_ | int | integer[Unknown] = ..., dtype: None = ..., *, like: _SupportsArrayFunc | None = ...) -> ndarray[Any, dtype[signedinteger[Any]]], (stop: bool | bool_ | int | integer[Unknown] | float | floating[Unknown], /, *, dtype: None = ..., like: _SupportsArrayFunc | None = ...) -> ndarray[Any, dtype[floating[Any]]], (start: bool | bool_ | int | integer[Unknown] | float | floating[Unknown], stop: bool | bool_ | int | integer[Unknown] | float | floating[Unknown], step: bool | bool_ | int | integer[Unknown] | float | floating[Unknown] = ..., dtype: None = ..., *, like: _SupportsArrayFunc | None = ...) -> ndarray[Any, dtype[floating[Any]]], (stop: bool | bool_ | int | integer[Unknown] | timedelta64, /, *, dtype: None = ..., like: _SupportsArrayFunc | None = ...) -> ndarray[Any, dtype[timedelta64]], (start: bool | bool_ | int | integer[Unknown] | timedelta64, stop: bool | bool_ | int | integer[Unknown] | timedelta64, step: bool | bool_ | int | integer[Unknown] | timedelta64 = ..., dtype: None = ..., *, like: _SupportsArrayFunc | None = ...) -> ndarray[Any, dtype[timedelta64]], (start: datetime64, stop: datetime64, step: datetime64 = ..., dtype: None = ..., *, like: _SupportsArrayFunc | None = ...) -> ndarray[Any, dtype[datetime64]], (stop: Any, /, *, dtype: dtype[_SCT@arange] | Type[_SCT@arange] | _SupportsDType[dtype[_SCT@arange]], like: _SupportsArrayFunc | None = ...) -> ndarray[Any, dtype[_SCT@arange]], (start: Any, stop: Any, step: Any = ..., dtype: dtype[_SCT@arange] | Type[_SCT@arange] | _SupportsDType[dtype[_SCT@arange]] = ..., *, like: _SupportsArrayFunc | None = ...) -> ndarray[Any, dtype[_SCT@arange]], (stop: Any, /, *, dtype: dtype[Any] | type | _SupportsDType[dtype[Any]] | str | Tuple[Any, int] | Tuple[Any, SupportsIndex | Sequence[SupportsIndex]] | List[Any] | _DTypeDict | Tuple[Any, Any] | None, like: _SupportsArrayFunc | None = ...) -> ndarray[Any, dtype[Any]], (start: Any, stop: Any, step: Any = ..., dtype: dtype[Any] | type | _SupportsDType[dtype[Any]] | str | Tuple[Any, int] | Tuple[Any, SupportsIndex | Sequence[SupportsIndex]] | List[Any] | _DTypeDict | Tuple[Any, Any] | None = ..., *, like: _SupportsArrayFunc | None = ...) -> ndarray[Any, dtype[Any]]]" (reportUnknownMemberType)
```

With the `reveal_type` returning: `Type of "numpy.arange(5)" is "ndarray[Any, dtype[signedinteger[Any]]]"`

Seeing how one of the `arange` overloads supports `Any` as the arguments you would expect at least that version to match, but it doesn't either.

What am I missing and how can I make it work without resorting to ignoring rules?

Related issue: https://github.com/microsoft/pyright/discussions/3455

**Comment by erictraut** [comment]
Using strict mode with libraries whose types are incomplete is going to be fraught with problems. Unfortunately, numpy is one such library.

Your options are:
1. Disable strict type checking when using numpy or other libraries that have missing type information. You can enable a subset of the strict-mode diagnostic checks but leave the `reportUnknown*` diagnostic checks disabled. For a full list of diagnostic checks and their default severity settings for "basic" and "strict" mode type checking, refer to [this table](https://microsoft.github.io/pyright/#/configuration?id=diagnostic-rule-defaults). 
2. Roll your own type stubs for the subset of numpy symbols that your project uses. You can then supply complete type information for these symbols. This is the approach I've used in the past for my team's code base, but it was significant work to create and maintain these type stubs, and the resulting stubs wouldn't work with other projects because they covered only the small subset of symbols that my team used from numpy.
3. Suppress individual errors using `# type: ignore` or `# pyright: ignore[reportUnknownMemberType]` comments.
4. Contribute type improvements to numpy through PRs . This doesn't address your problem in the short term, but it improves the situation for the entire Python community over time.

**Reply by wolph** [reply]
Thank you for the great and very quick response!
In that case I'm going to go for a combination between 2 and 4 probably. Assuming the numpy team is even interested in these types of pull requests that is :)

**Comment by wolph** [comment, marked as answer]
Upon further inspection it seems that numpy 1.25 has a fix for the issue: https://github.com/numpy/numpy/commit/1a4a2adffb5ebc0f6b12a77a940bfcb2ebaa86bb

The issue was apparently that these types were lacking the generic arguments.
I really wish pyright had a clearer error for this because this was quite hard to track down and ultimately useless since it's already been fixed apparently.
